modules/heat_conduction/src/meshgenerators/PatchSidesetGenerator.C

```diff
@@ -0,0 +1,286 @@
+      "n_patches", "n_patches>0", "Number of patches");
+
+  MooseEnum partitioning("default=-3 metis=-2 parmetis=-1 linear=0 centroid hilbert_sfc morton_sfc",
+                         "default");
+  params.addParam<MooseEnum>(
+      "partitioner",
+      partitioning,
+      "Specifies a mesh partitioner to use when splitting the mesh for a parallel computation.");
+  MooseEnum direction("x y z radial");
+  params.addParam<MooseEnum>("centroid_partitioner_direction",
+                             direction,
+                             "Specifies the sort direction if using the centroid partitioner. "
+                             "Available options: x, y, z, radial");
+
+  params.addParamNamesToGroup("partitioner centroid_partitioner_direction", "Partitioning");
+
+  params.addClassDescription(
+      "Divides the given sideset into smaller patches of roughly equal size.");
+
+  return params;
+}
+
+PatchSidesetGenerator::PatchSidesetGenerator(const InputParameters & parameters)
+  : MeshGenerator(parameters),
+    _input(getMesh("input")),
+    _n_patches(getParam<unsigned int>("n_patches")),
+    _sideset(getParam<boundary_id_type>("sideset")),
+    _partitioner_name(getParam<MooseEnum>("partitioner"))
+{
+}
+
+std::unique_ptr<MeshBase>
+PatchSidesetGenerator::generate()
+{
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+
+  bool distributed = false;
+  if (typeid(mesh).name() == typeid(std::unique_ptr<DistributedMesh>).name())
+    distributed = true;
+
+  if (distributed)
+    mooseWarning("Distributed mesh may not work yet, because communication is not done.");
+
+  // Get a reference to our BoundaryInfo object for later use
+  BoundaryInfo & boundary_info = mesh->get_boundary_info();
+
+  // get a list of all sides; vector of tuples (elem, loc_side, side_set)
+  auto side_list = boundary_info.build_active_side_list();
+
+  // create a dim - 1 dimensional mesh
+  auto boundary_mesh =
+      libmesh_make_unique<libMesh::ReplicatedMesh>(comm(), mesh->mesh_dimension() - 1);
+  boundary_mesh->set_mesh_dimension(mesh->mesh_dimension() - 1);
+  boundary_mesh->set_spatial_dimension(mesh->mesh_dimension());
+
+  // nodes in the new mesh by boundary_node_id (index)
+  std::vector<Node *> boundary_nodes;
+  // a map from the node numbering on the volumetric mesh to the numbering
+  // on the boundary_mesh
+  std::map<dof_id_type, dof_id_type> mesh_node_id_to_boundary_node_id;
+  // a local counter keeping track of how many entries have been added to boundary_nodes
+  dof_id_type boundary_node_id = 0;
+  // a map from new element id in the boundary mesh to the element id/side/sideset
+  // tuple it came from
+  std::map<dof_id_type, std::tuple<dof_id_type, unsigned short int, boundary_id_type>>
+      boundary_elem_to_mesh_elem;
+  for (auto & side : side_list)
+  {
+    if (std::get<2>(side) == _sideset)
+    {
+      // the original volumetric mesh element
+      const Elem * elem = mesh->elem_ptr(std::get<0>(side));
```

Comment:
Yeah - this is likely to fail on distributed mesh. Probably should just turn on the error for now.